File: bench/rust/src/scheme/record.rs
use super::{env_define, EnvRef, EvalError, Expr, Value};
use std::cell::RefCell;
use std::rc::Rc;

pub(super) type RecordRef = Rc<RecordInstance>;
type RecordTypeRef = Rc<RecordType>;

struct RecordType {
    name: String,
    field_count: usize,
}

pub(super) struct RecordInstance {
    record_type: RecordTypeRef,
    fields: RefCell<Vec<Value>>,
}

pub(super) struct NativeProcedure {
    name: String,
    kind: NativeProcedureKind,
}

enum NativeProcedureKind {
    Constructor {
        record_type: RecordTypeRef,
    },
    Predicate {
        record_type: RecordTypeRef,
    },
    Accessor {
        record_type: RecordTypeRef,
        field_index: usize,
    },
}

impl NativeProcedure {
    fn new(name: String, kind: NativeProcedureKind) -> Self {
        Self { name, kind }
    }

    pub(super) fn apply(&self, args: &[Value]) -> Result<Value, EvalError> {
        match &self.kind {
            NativeProcedureKind::Constructor { record_type } => {
                if args.len() != record_type.field_count {
                    return Err(EvalError::WrongArgCount {
                        name: self.name.clone(),
                        expected: format!("exactly {} argument(s)", record_type.field_count),
                        got: args.len(),
                    });
                }

                Ok(Value::Record(Rc::new(RecordInstance {
                    record_type: Rc::clone(record_type),
                    fields: RefCell::new(args.to_vec()),
                })))
            }
            NativeProcedureKind::Predicate { record_type } => {
                let [value] = args else {
                    return Err(EvalError::WrongArgCount {
                        name: self.name.clone(),
                        expected: "exactly 1 argument".into(),
                        got: args.len(),
                    });
                };

                Ok(Value::Boolean(matches!(
                    value,
                    Value::Record(record) if Rc::ptr_eq(&record.record_type, record_type)
                )))
            }
            NativeProcedureKind::Accessor {
                record_type,
                field_index,
            } => {
                let [value] = args else {
                    return Err(EvalError::WrongArgCount {
                        name: self.name.clone(),
                        expected: "exactly 1 argument".into(),
                        got: args.len(),
                    });
                };

                let Value::Record(record) = value else {
                    return Err(EvalError::RecordTypeMismatch {
                        expected: record_type.name.clone(),
                        found: value.type_name().into(),
                    });
                };

                if !Rc::ptr_eq(&record.record_type, record_type) {
                    return Err(EvalError::RecordTypeMismatch {
                        expected: record_type.name.clone(),
                        found: record.record_type.name.clone(),
                    });
                }

                Ok(record
                    .fields
                    .borrow()
                    .get(*field_index)
                    .cloned()
                    .expect("record accessor index must be valid"))
            }
        }
    }
}

pub(super) fn render_record(record: &RecordRef) -> String {
    format!("#<record {}>", record.record_type.name)
}

pub(super) fn define_record_type(args: &[Expr], env: &EnvRef) -> Result<Value, EvalError> {
    let [record_name, constructor_spec, predicate_name, field_specs @ ..] = args else {
        return Err(EvalError::SyntaxError {
            message: "invalid define-record-type".into(),
        });
    };

    let record_name = parse_record_symbol(record_name, "record type name must be a symbol")?;
    let (constructor_name, field_count) = parse_record_constructor_spec(constructor_spec)?;
    let predicate_name =
        parse_record_symbol(predicate_name, "record predicate name must be a symbol")?;
    let accessor_names = parse_record_field_specs(field_specs)?;

    if field_count != accessor_names.len() {
        return Err(EvalError::SyntaxError {
            message: "record constructor and field definitions must have matching arity".into(),
        });
    }

    let record_type = Rc::new(RecordType {
        name: record_name,
        field_count,
    });

    env_define(
        env,
        constructor_name.clone(),
        Value::NativeProcedure(Rc::new(NativeProcedure::new(
            constructor_name,
            NativeProcedureKind::Constructor {
                record_type: Rc::clone(&record_type),
            },
        ))),
    );

    env_define(
        env,
        predicate_name.clone(),
        Value::NativeProcedure(Rc::new(NativeProcedure::new(
            predicate_name,
            NativeProcedureKind::Predicate {
                record_type: Rc::clone(&record_type),
            },
        ))),
    );

    for (field_index, accessor_name) in accessor_names.into_iter().enumerate() {
        env_define(
            env,
            accessor_name.clone(),
            Value::NativeProcedure(Rc::new(NativeProcedure::new(
                accessor_name,
                NativeProcedureKind::Accessor {
                    record_type: Rc::clone(&record_type),
                    field_index,
                },
            ))),
        );
    }

    Ok(Value::Void)
}

fn parse_record_symbol(item: &Expr, message: &str) -> Result<String, EvalError> {
    match item {
        Expr::Symbol(name, _) => Ok(name.clone()),
        _ => Err(EvalError::SyntaxError {
            message: message.into(),
        }),
    }
}

fn parse_record_constructor_spec(item: &Expr) -> Result<(String, usize), EvalError> {
    let Expr::List(items, _) = item else {
        return Err(EvalError::SyntaxError {
            message: "record constructor spec must be a list".into(),
        });
    };

    let Some((name, fields)) = items.split_first() else {
        return Err(EvalError::SyntaxError {
            message: "record constructor spec cannot be empty".into(),
        });
    };

    let constructor_name = parse_record_symbol(name, "record constructor name must be a symbol")?;
    let fields = super::parse_required_param_names(fields)?;
    Ok((constructor_name, fields.len()))
}

fn parse_record_field_specs(field_specs: &[Expr]) -> Result<Vec<String>, EvalError> {
    let mut accessors = Vec::with_capacity(field_specs.len());
    for field in field_specs {
        let Expr::List(items, _) = field else {
            return Err(EvalError::SyntaxError {
                message: "record field spec must be a list".into(),
            });
        };

        let [_, accessor] = items.as_slice() else {
            return Err(EvalError::SyntaxError {
                message: "record field spec must be (field accessor)".into(),
            });
        };

        accessors.push(parse_record_symbol(
            accessor,
            "record accessor name must be a symbol",
        )?);
    }

    Ok(accessors)
}

File: bench/go/record.go
package ming

import "fmt"

type recordType struct {
	name   string
	fields []recordFieldSpec
}

type recordFieldSpec struct {
	name        symbolExpr
	accessor    symbolExpr
	hasAccessor bool
	mutator     symbolExpr
	hasMutator  bool
}

type recordValue struct {
	typ    *recordType
	fields []any
}

func evalDefineRecordType(scope *env, args []any) (any, error) {
	if len(args) < 3 {
		return nil, &EvalError{Message: "define-record-type expects a type name, constructor, predicate, and fields"}
	}

	typeName, ok := args[0].(symbolExpr)
	if !ok {
		return nil, exprSourcePos(args[0]).errorf("define-record-type requires a type name")
	}

	constructorForm, ok := args[1].(listExpr)
	if !ok || len(constructorForm.elements) == 0 {
		return nil, exprSourcePos(args[1]).errorf("define-record-type constructor must be a list")
	}

	constructorName, ok := constructorForm.elements[0].(symbolExpr)
	if !ok {
		return nil, exprSourcePos(constructorForm.elements[0]).errorf("define-record-type constructor name must be a symbol")
	}

	constructorParams := constructorForm.elements[1:]
	for _, paramExpr := range constructorParams {
		if _, ok := paramExpr.(symbolExpr); !ok {
			return nil, exprSourcePos(paramExpr).errorf("define-record-type constructor parameters must be symbols")
		}
	}

	predicateName, ok := args[2].(symbolExpr)
	if !ok {
		return nil, exprSourcePos(args[2]).errorf("define-record-type predicate must be a symbol")
	}

	fields := make([]recordFieldSpec, len(args)-3)
	for i, fieldExpr := range args[3:] {
		field, err := parseRecordFieldSpec(fieldExpr)
		if err != nil {
			return nil, err
		}
		fields[i] = field
	}

	if len(constructorParams) != len(fields) {
		return nil, constructorForm.pos.errorf("define-record-type constructor arity does not match field count")
	}

	recordType := &recordType{
		name:   typeName.name,
		fields: fields,
	}

	fieldCount := len(fields)
	scope.defineSymbol(constructorName, builtinProc{
		name: constructorName.name,
		fn: func(args []any) (any, error) {
			if len(args) != fieldCount {
				return nil, &EvalError{Message: fmt.Sprintf("%s expects exactly %d arguments", constructorName.name, fieldCount)}
			}

			values := append([]any(nil), args...)
			return &recordValue{
				typ:    recordType,
				fields: values,
			}, nil
		},
	})

	scope.defineSymbol(predicateName, builtinProc{
		name: predicateName.name,
		fn: func(args []any) (any, error) {
			if len(args) != 1 {
				return nil, &EvalError{Message: fmt.Sprintf("%s expects exactly 1 argument", predicateName.name)}
			}

			record, ok := args[0].(*recordValue)
			return ok && record.typ == recordType, nil
		},
	})

	for index, field := range fields {
		index := index
		field := field

		if field.hasAccessor {
			scope.defineSymbol(field.accessor, builtinProc{
				name: field.accessor.name,
				fn: func(args []any) (any, error) {
					if len(args) != 1 {
						return nil, &EvalError{Message: fmt.Sprintf("%s expects exactly 1 argument", field.accessor.name)}
					}

					record, err := expectRecordOfType(args[0], recordType, field.accessor.name)
					if err != nil {
						return nil, err
					}
					return record.fields[index], nil
				},
			})
		}

		if field.hasMutator {
			scope.defineSymbol(field.mutator, builtinProc{
				name: field.mutator.name,
				fn: func(args []any) (any, error) {
					if len(args) != 2 {
						return nil, &EvalError{Message: fmt.Sprintf("%s expects exactly 2 arguments", field.mutator.name)}
					}

					record, err := expectRecordOfType(args[0], recordType, field.mutator.name)
					if err != nil {
						return nil, err
					}

					record.fields[index] = args[1]
					return voidValue{}, nil
				},
			})
		}
	}

	return voidValue{}, nil
}

func parseRecordFieldSpec(expr any) (recordFieldSpec, error) {
	switch field := expr.(type) {
	case symbolExpr:
		return recordFieldSpec{name: field}, nil
	case listExpr:
		if len(field.elements) == 0 || len(field.elements) > 3 {
			return recordFieldSpec{}, field.pos.errorf("record field spec must contain 1 to 3 identifiers")
		}

		name, ok := field.elements[0].(symbolExpr)
		if !ok {
			return recordFieldSpec{}, exprSourcePos(field.elements[0]).errorf("record field name must be a symbol")
		}

		spec := recordFieldSpec{name: name}
		if len(field.elements) >= 2 {
			accessor, ok := field.elements[1].(symbolExpr)
			if !ok {
				return recordFieldSpec{}, exprSourcePos(field.elements[1]).errorf("record accessor name must be a symbol")
			}
			spec.accessor = accessor
			spec.hasAccessor = true
		}

		if len(field.elements) == 3 {
			mutator, ok := field.elements[2].(symbolExpr)
			if !ok {
				return recordFieldSpec{}, exprSourcePos(field.elements[2]).errorf("record mutator name must be a symbol")
			}
			spec.mutator = mutator
			spec.hasMutator = true
		}

		return spec, nil
	default:
		return recordFieldSpec{}, exprSourcePos(expr).errorf("record field spec must be a symbol or list")
	}
}

func expectRecordOfType(value any, recordType *recordType, who string) (*recordValue, error) {
	record, ok := value.(*recordValue)
	if !ok || record.typ != recordType {
		return nil, &EvalError{Message: fmt.Sprintf("%s expects a %s record, got %s", who, recordType.name, typeName(value))}
	}
	return record, nil
}
